### PR TITLE
Generic defaults

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
A mostly-complete implementation for #2175

Not done yet:
- [ ] Check that all defaults, when present, are identical
- [ ] Check that defaults are assignable to constraints

/cc @ahejlsberg

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2948,10 +2948,25 @@ namespace ts {
             return getClassExtendsHeritageClauseElement(<ClassLikeDeclaration>type.symbol.valueDeclaration);
         }
 
+        function getRequiredTypeArgumentCount(typeParameters: TypeParameter[]) {
+            if (typeParameters) {
+                for (let i = typeParameters.length - 1; i >= 0; i--) {
+                    if (!typeParameters[i].default) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: This isn't ES3-compliant and we don't downlevel this as `typeParameters[i]["default"]`.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -2948,10 +2948,25 @@ namespace ts {
             return getClassExtendsHeritageClauseElement(<ClassLikeDeclaration>type.symbol.valueDeclaration);
         }
 
+        function getRequiredTypeArgumentCount(typeParameters: TypeParameter[]) {
+            if (typeParameters) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: You can negate this condition to reduce nesting.

File: tests/cases/conformance/types/typeParameters/typeParameterDefaultsErrors.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+interface HasDefault2<T1, T2 = number> {
+	x1: T1;
+	x2: T2;
+}
+// Still an error
+let a4: HasDefault2;
+
+/** Required may not follow optional */
+interface Wrong<T, U = number> { }
```
Comment (mhegazy) [inline tests/cases/conformance/types/typeParameters/typeParameterDefaultsErrors.ts]: not an error.

**Comment by mhegazy** [comment]
@RyanCavanaugh can you please refresh this PR?

File: tests/baselines/reference/typeParameterDefaultsErrors.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+interface HasDefault2<T1, T2 = number> {
+	x1: T1;
+	x2: T2;
+}
+// Still an error
+let a4: HasDefault2;
+
+/** Required may not follow optional */
+interface Wrong<T, U = number> { }
```
Comment (isiahmeadows) [inline tests/baselines/reference/typeParameterDefaultsErrors.js]: Did you mean `interface Wrong<T = number, U> {}`?

**Comment by RyanCavanaugh** [comment]
Superceded by #13487

**Pull request closed without merging**